tests/_utils/test_functions.py

```diff
@@ -0,0 +1,1135 @@
+
+    @property
+    def message_param(self) -> Any:
+        """Returns the assistants's response as a message parameter."""
+        pass
+
+    @computed_field
+    @property
+    def tools(self) -> list[BaseTool[Any]] | None:
+        """Returns any available tool calls as their `OpenAITool` definition.
+
+        Raises:
+            ValidationError: if a tool call doesn't match the tool's schema.
+            ValueError: if the model refused to response, in which case the error
+                message will be the refusal.
+        """
+        if hasattr(self.response.choices[0].message, "refusal") and (
+            refusal := self.response.choices[0].message.refusal
+        ):
+            raise ValueError(refusal)
+
+        tool_calls = self.response.choices[0].message.tool_calls
+        if not self.tool_types or not tool_calls:
+            return None
+
+        extracted_tools = []
+        for tool_call in tool_calls:
+            for tool_type in self.tool_types:
+                if tool_call.function.name == tool_type._name():
+                    extracted_tools.append(tool_type.from_tool_call(tool_call))  # pyright: ignore [reportAttributeAccessIssue]
+                    break
+
+        return extracted_tools
+
+    @property
+    def tool(self) -> _BaseToolT | None:  # pyright: ignore [reportInvalidTypeVarUse]
+        """Returns the tool that was used to generate the response."""
+        pass
+
+    @classmethod
+    def tool_message_params(
+        cls, tools_and_outputs: list[tuple[BaseTool[Any], str]]
+    ) -> list[ChatCompletionToolMessageParam]:
+        """Returns the tool message parameters for tool call results.
+
+        Args:
+            tools_and_outputs: The list of tools and their outputs from which the tool
+                message parameters should be constructed.
+
+        Returns:
+            The list of constructed `ChatCompletionToolMessageParam` parameters.
+        """
+        return [
+            ChatCompletionToolMessageParam(  # pyright: ignore [reportCallIssue]
+                role="tool",
+                content=output,
+                tool_call_id=tool.tool_call.id,  # pyright: ignore [reportAttributeAccessIssue]
+                name=tool._name(),  # pyright: ignore [reportCallIssue]
+            )
+            for tool, output in tools_and_outputs
+        ]
+
+    ...
+
+
+@pytest.fixture
+def mock_chat_completion() -> ChatCompletion:
+    """Create a mock ChatCompletion"""
+    return ChatCompletion(
+        id="test-id",
+        choices=[
+            Choice(
+                finish_reason="stop",
+                index=0,
+                message=ChatCompletionMessage(
+                    content="test content",
+                    role="assistant",
+                ),
+            )
+        ],
+        created=0,
+        model="gpt-4",
+        object="chat.completion",
+        system_fingerprint="test-fingerprint",
+    )
+
+
+class MockBaseTool(BaseTool[Any], ABC):
```

Comment:
same thing here and for all of these types (which i'm fairly certain we already defined), I think for instances of mocks like this we should pull them out into a conftest if they are used in multiple test locations